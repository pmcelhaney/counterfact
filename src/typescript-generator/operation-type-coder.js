import nodePath from "node:path";

import { Coder } from "./coder.js";
import { SchemaTypeCoder } from "./schema-type-coder.js";
import { ParametersTypeCoder } from "./parameters-type-coder.js";
import { ResponseTypeCoder } from "./response-type-coder.js";
import { ContextCoder } from "./context-coder.js";

export class OperationTypeCoder extends Coder {
  names() {
    return super.names(
      `HTTP_${this.requirement.url.split("/").at(-1).toUpperCase()}`
    );
  }

  responseTypes(script) {
    return this.requirement
      .get("responses")
      .flatMap((response, responseCode) => {
        const status =
          responseCode === "default"
            ? "number | undefined"
            : Number.parseInt(responseCode, 10);

        if (response.has("content")) {
          return response.get("content").map(
            (content, contentType) => `{  
              status: ${status}, 
              contentType?: "${contentType}",
              body?: ${new SchemaTypeCoder(content.get("schema")).write(script)}
            }`
          );
        }

        if (response.has("schema")) {
          return this.requirement
            .get("produces")
            .data.map(
              (contentType) => `{
            status: ${status},
            contentType?: "${contentType}",
            body?: ${new SchemaTypeCoder(response.get("schema")).write(script)}
          }`
            )
            .join(" | ");
        }

        return `{  
          status: ${status} 
        }`;
      })

      .join(" | ");
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath.join("path-types", pathString)}.types.ts`;
  }

  write(script) {
    const contextImportName = script.importDefault(
      new ContextCoder(this.requirement)
    );

    const parameters = this.requirement.get("parameters");

    const queryType =
      parameters === undefined
        ? "never"
        : new ParametersTypeCoder(parameters, "query").write(script);

    const pathType =
      parameters === undefined
        ? "never"
        : new ParametersTypeCoder(parameters, "path").write(script);

    const headerType =
      parameters === undefined
        ? "never"
        : new ParametersTypeCoder(parameters, "header").write(script);

    const bodyRequirement = this.requirement.get("consumes")
      ? parameters
          .find((parameter) =>
            ["body", "formData"].includes(parameter.get("in").data)
          )
          .get("schema")
      : this.requirement.select("requestBody/content/application~1json/schema");

    const bodyType = bodyRequirement
      ? new SchemaTypeCoder(bodyRequirement).write(script)
      : "undefined";

    const responseType = new ResponseTypeCoder(
      this.requirement.get("responses"),
      this.requirement.get("produces")?.data
    ).write(script);

    return `({ query, path, header, body, context }: { query: ${queryType}, path: ${pathType}, header: ${headerType}, body: ${bodyType}, context: typeof ${contextImportName}, response: ${responseType} }) => ${this.responseTypes(
      script
    )} | { status: 415, contentType: "text/plain", body: string }
    | void`;
  }
}
