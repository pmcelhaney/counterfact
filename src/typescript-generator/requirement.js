export class Requirement {
  constructor(data, url = "", specification = undefined) {
    this.data = data;
    this.url = url;
    this.specification = specification;
  }

  get isReference() {
    return this.data?.$ref !== undefined;
  }

  reference() {
    return this.specification.requirementAt(this.data.$ref, this.url);
  }

  has(item) {
    return item in this.data;
  }

  get(item) {
    if (!this.has(item)) {
      return undefined;
    }

    return new Requirement(
      this.data[item],
      `${this.url}/${item}`,
      this.specification,
    );
  }

  select(path, data = this.data, basePath = "") {
    const [head, ...tail] = path.split("/");

    const branch = data[this.unescapeJsonPointer(head)];

    if (!branch) {
      return undefined;
    }

    if (tail.length === 0) {
      return new Requirement(
        branch,
        `${this.url}/${basePath}${head}`,
        this.specification,
      );
    }

    return this.select(tail.join("/"), branch, `${basePath}${head}/`);
  }

  forEach(callback) {
    Object.keys(this.data).forEach((key) => {
      callback(this.select(this.escapeJsonPointer(key)), key);
    });
  }

  map(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach((value, key) => result.push(callback(value, key)));

    return result;
  }

  flatMap(callback) {
    // eslint-disable-next-line unicorn/prefer-array-flat-map
    return this.map(callback).flat();
  }

  find(callback) {
    // eslint-disable-next-line init-declarations
    let result;

    this.forEach((value, key) => {
      // eslint-disable-next-line  n/callback-return
      if (result === undefined && callback(value, key)) {
        result = value;
      }
    });

    return result;
  }

  escapeJsonPointer(string) {
    return string.replaceAll("~", "~0").replaceAll("/", "~1");
  }

  unescapeJsonPointer(pointer) {
    return pointer.replaceAll("~1", "/").replaceAll("~0", "~");
  }
}
