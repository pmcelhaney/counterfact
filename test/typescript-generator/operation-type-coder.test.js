/* eslint-disable jest/no-restricted-matchers */
import prettier from "prettier";

import { OperationTypeCoder } from "../../src/typescript-generator/operation-type-coder.js";
import { Requirement } from "../../src/typescript-generator/requirement.js";

function format(code) {
  return prettier.format(code, { parser: "typescript" });
}

const dummyScript = {
  path: ".",

  importDefault() {
    return "default";
  },

  importExternalType() {
    return "ExternalType";
  },

  importType() {
    return "Type";
  },

  import() {
    return "schema";
  },
};

describe("an OperationTypeCoder", () => {
  it("generates a list of potential names", () => {
    const coder = new OperationTypeCoder(
      new Requirement({}, "#/paths/hello/get")
    );

    const [one, two, three] = coder.names();

    expect([one, two, three]).toStrictEqual([
      "HTTP_GET",
      "HTTP_GET2",
      "HTTP_GET3",
    ]);
  });

  it("creates a type declaration", () => {
    const coder = new OperationTypeCoder(
      new Requirement({}, "#/paths/hello/get")
    );

    expect(coder.typeDeclaration(undefined, dummyScript)).toBe("");
  });

  it("returns the module path", () => {
    const coder = new OperationTypeCoder(
      new Requirement({}, "#/paths/hello~1world/get")
    );

    expect(coder.modulePath()).toBe("path-types/hello/world.types.ts");
  });

  it("generates a complex post operation", () => {
    const requirement = new Requirement(
      {
        parameters: [
          { name: "id", in: "path", schema: { type: "string" } },
          { name: "name", in: "query", schema: { type: "string" } },
          { name: "name", in: "header", schema: { type: "string" } },
        ],

        requestBody: {
          content: {
            "application/json": {
              schema: { $ref: "#/components/schemas/Example" },
            },
          },
        },

        responses: {
          200: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Example" },

                examples: {
                  "first-example": {
                    value: "first",
                  },

                  "second-example": {
                    value: "second",
                  },
                },
              },
            },
          },

          400: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Error" },
              },
            },
          },

          default: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Error" },
              },
            },
          },
        },
      },
      "#/paths/hello/post"
    );

    const coder = new OperationTypeCoder(requirement);

    expect(
      format(`type TestType = ${coder.write(dummyScript)}`)
    ).toMatchSnapshot();
  });

  it("generates a complex post operation (OpenAPI 2)", () => {
    const requirement = new Requirement(
      {
        consumes: ["application/json"],
        produces: ["application/json"],

        parameters: [
          { name: "id", in: "path", type: "string" },
          { name: "name", in: "query", type: "string" },
          { name: "name", in: "header", type: "string" },
          {
            name: "body",
            in: "body",
            schema: { $ref: "#/components/schemas/Example" },
          },
        ],

        responses: {
          200: {
            schema: { $ref: "#/components/schemas/Example" },
          },

          400: {
            schema: { $ref: "#/components/schemas/Error" },
          },

          default: {
            schema: { $ref: "#/components/schemas/Error" },
          },
        },
      },
      "#/paths/hello/post"
    );

    const coder = new OperationTypeCoder(requirement);

    expect(
      format(`type TestType = ${coder.write(dummyScript)}`)
    ).toMatchSnapshot();
  });

  it("generates a simple get operation", () => {
    const requirement = new Requirement(
      {
        responses: {
          default: {
            content: {
              "application/json": {
                schema: { $ref: "#/components/schemas/Example" },
              },
            },
          },
        },
      },
      "#/paths/hello/get"
    );

    const coder = new OperationTypeCoder(requirement);

    expect(
      format(`type TestType =${coder.write(dummyScript)}`)
    ).toMatchSnapshot();
  });

  it("generates a simple get operation (OpenAPI 2)", () => {
    const requirement = new Requirement(
      {
        produces: ["application/json"],

        responses: {
          default: {
            schema: { $ref: "#/components/schemas/Example" },
          },
        },
      },
      "#/paths/hello/get"
    );

    const coder = new OperationTypeCoder(requirement);

    expect(
      format(`type TestType =${coder.write(dummyScript)}`)
    ).toMatchSnapshot();
  });
});
