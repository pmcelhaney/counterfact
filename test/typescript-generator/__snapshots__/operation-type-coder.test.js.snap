// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`an OperationTypeCoder generates a complex post operation (OpenAPI 2) 1`] = `
"type TestType = ({
  query,
  path,
  header,
  body,
  context,
}: {
  query: { name?: string };
  path: { id?: string };
  header: { name?: string };
  body: Type;
  context: typeof default;
  response: ResponseBuilderFactory<{
    200: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
    400: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
    [statusCode in Exclude<HttpStatusCode, 200 | 400>]: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
  }>;
}) =>
  | {
      status: 200;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | {
      status: 400;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | {
      status: number | undefined;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | { status: 415; contentType: \\"text/plain\\"; body: string }
  | void;
"
`;

exports[`an OperationTypeCoder generates a complex post operation 1`] = `
"type TestType = ({
  query,
  path,
  header,
  body,
  context,
}: {
  query: { name?: string };
  path: { id?: string };
  header: { name?: string };
  body: Type;
  context: typeof default;
  response: ResponseBuilderFactory<{
    200: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
    400: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
    [statusCode in Exclude<HttpStatusCode, 200 | 400>]: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
  }>;
}) =>
  | {
      status: 200;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | {
      status: 400;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | {
      status: number | undefined;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | { status: 415; contentType: \\"text/plain\\"; body: string }
  | void;
"
`;

exports[`an OperationTypeCoder generates a simple get operation (OpenAPI 2) 1`] = `
"type TestType = ({
  query,
  path,
  header,
  body,
  context,
}: {
  query: never;
  path: never;
  header: never;
  body: undefined;
  context: typeof default;
  response: ResponseBuilderFactory<{
    [statusCode in HttpStatusCode]: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
  }>;
}) =>
  | {
      status: number | undefined;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | { status: 415; contentType: \\"text/plain\\"; body: string }
  | void;
"
`;

exports[`an OperationTypeCoder generates a simple get operation 1`] = `
"type TestType = ({
  query,
  path,
  header,
  body,
  context,
}: {
  query: never;
  path: never;
  header: never;
  body: undefined;
  context: typeof default;
  response: ResponseBuilderFactory<{
    [statusCode in HttpStatusCode]: {
      headers: {};
      content: {
        \\"application/json\\": {
          schema: Type;
        };
      };
    };
  }>;
}) =>
  | {
      status: number | undefined;
      contentType?: \\"application/json\\";
      body?: Type;
    }
  | { status: 415; contentType: \\"text/plain\\"; body: string }
  | void;
"
`;
